host/class/uvc/usb_host_uvc_2/uvc_bulk.c

```diff
@@ -0,0 +1,199 @@
+/*
+ * SPDX-FileCopyrightText: 2024 Espressif Systems (Shanghai) CO LTD
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdbool.h>
+#include <string.h> // For memcpy
+
+#include "esp_log.h"
+
+#include "uvc_types_priv.h"
+#include "uvc_check_priv.h"
+#include "uvc_frame_priv.h"
+#include "uvc_critical_priv.h"
+
+static const char *TAG = "uvc-bulk";
+
+/**
+ * @brief Callback function for handling Bulk USB transfers from a UVC camera.
+ *
+ * This function processes Bulk transfer packets, managing frame data according to a state machine. The characteristics
+ * and requirements of Bulk transfers are as follows:
+ *
+ * - **CRC Included**: Ensures no errors in frame data.
+ * - **ACK Mechanism**: Missed packets are retransmitted, ensuring reliable data delivery.
+ * - **Packet Headers**: Not all packets contain headers; the end of a transfer is indicated by a short packet
+ *   (less than the maximum packet size), and the next packet usually contains a header.
+ *
+ * To process these packets, a state machine is implemented to track the next expected Bulk packet type:
+ * - Start of Frame (SoF), usually followed by data packets in the same transfer
+ * - Data packets (no header)
+ * - End of Frame (EoF)
+ *
+ * The function handles USB transfer statuses, manages frame buffers, and invokes user-defined callbacks for
+ * completed frames.
+ *
+ * @param[in] transfer Pointer to the completed USB transfer structure.
+ */
+void bulk_transfer_callback(usb_transfer_t *transfer)
+{
+    ESP_LOGD(TAG, "%s", __FUNCTION__);
+    uvc_stream_t *uvc_stream = (uvc_stream_t *)transfer->context;
+
+    UVC_ENTER_CRITICAL();
+    bool streaming_on = uvc_stream->streaming;
+    UVC_EXIT_CRITICAL();
+    if (!streaming_on) {
+        return; // If the streaming was turned off, we don't have to do anything
+    }
+
+    // In BULK implementation, 'payload' is a constant pointer to constant data,
+    // meaning both the pointer and the data it points to cannot be changed.
+    // This contrasts with the ISOC implementation, where 'payload' is a variable
+    // pointer and is increased after every ISOC packet processing
+    const uint8_t *const payload = transfer->data_buffer;
+    const uint8_t *payload_data  = payload;
+    size_t payload_data_len      = transfer->actual_num_bytes;
+
+    // Check USB transfer status
+    switch (transfer->status) {
+    case USB_TRANSFER_STATUS_COMPLETED:
+        break;
+    case USB_TRANSFER_STATUS_NO_DEVICE:
+    case USB_TRANSFER_STATUS_CANCELED:
+    case USB_TRANSFER_STATUS_ERROR:
+    case USB_TRANSFER_STATUS_OVERFLOW:
+    case USB_TRANSFER_STATUS_STALL:
+        // On Bulk errors we stop the stream
+        //@todo not tested yet
+        //@todo Stall, error and overflow errors should be propagated to the user
```

Comment:
Is the STALL is a regular logic? 
Like STALLing the unknown control transfers in general? It is not the problem, device just got the unknown request. 

Is it the type of an error we need to stop stream after?

Reply:
I have not encoutered STALL during my tests. The BULK IN transfer should never be STALLed, so we might as well just `abort` here